diff -U 4 -H -d -p -r -N -- a/osgi/framework/src/j3/vm/OSGi.java b/osgi/framework/src/j3/vm/OSGi.java
--- a/osgi/framework/src/j3/vm/OSGi.java	1970-01-01 01:00:00.000000000 +0100
+++ b/osgi/framework/src/j3/vm/OSGi.java	2013-07-03 17:15:27.382558975 +0200
@@ -0,0 +1,6 @@
+package j3.vm;
+
+public class OSGi
+{
+	public static native void setBundleClassLoader(long bundleID, ClassLoader loaderObject);
+}
diff -U 4 -H -d -p -r -N -- a/osgi/framework/src/org/knopflerfish/framework/BundleGeneration.java b/osgi/framework/src/org/knopflerfish/framework/BundleGeneration.java
--- a/osgi/framework/src/org/knopflerfish/framework/BundleGeneration.java	2013-07-03 17:24:51.446553603 +0200
+++ b/osgi/framework/src/org/knopflerfish/framework/BundleGeneration.java	2013-07-03 17:17:48.950557625 +0200
@@ -162,8 +162,9 @@ public class BundleGeneration implements
     lazyExcludes = null;
     timeStamp = System.currentTimeMillis();
     bpkgs = new BundlePackages(this, exportStr);
     classLoader = b.getClassLoader();
+    j3.vm.OSGi.setBundleClassLoader(bundle.getBundleId(), classLoader);
   }
 
 
   /**
@@ -440,8 +441,9 @@ public class BundleGeneration implements
             }
           }
 	}
         classLoader = bundle.secure.newBundleClassLoader(this);
+        j3.vm.OSGi.setBundleClassLoader(bundle.getBundleId(), classLoader);
 
         return true;
       }
       if (fragments != null) {
